src/main/scheduler/scheduler.c

```diff
@@ -260,21 +265,34 @@ inline static timeUs_t getPeriodCalculationBasis(const cfTask_t* task)
     return *(timeUs_t*)((uint8_t*)task + periodCalculationBasisOffset);
 }
 
+extern FAST_CODE void taskMainPidLoop(timeUs_t currentTimeUs); //SCEDEBUG
+
 FAST_CODE void scheduler(void)
 {
     // Cache currentTime
     const timeUs_t currentTimeUs = micros();
+    timeUs_t remainingTime = UINT_MAX;
+    static timeUs_t maxRemainingTime = 0;
+    static timeUs_t movingSumRemainingTime = 0;
 
     // Check for realtime tasks
-    bool outsideRealtimeGuardInterval = true;
     for (const cfTask_t *task = queueFirst(); task != NULL && task->staticPriority >= TASK_PRIORITY_REALTIME; task = queueNext()) {
-        const timeUs_t nextExecuteAt = getPeriodCalculationBasis(task) + task->desiredPeriod;
-        if ((timeDelta_t)(currentTimeUs - nextExecuteAt) >= 0) {
-            outsideRealtimeGuardInterval = false;
-            break;
+        if (task->lastExecutedAt) {
+            timeUs_t taskDueTime = getPeriodCalculationBasis(task) + task->desiredPeriod;
+            timeUs_t taskRemainingTime = (taskDueTime > currentTimeUs) ? (taskDueTime - currentTimeUs) : 0;
+            if (taskRemainingTime < remainingTime) {
+                remainingTime = taskRemainingTime;
+
+                if (remainingTime > maxRemainingTime) {
+                    maxRemainingTime = remainingTime;
+                }
+            }
+            movingSumRemainingTime += remainingTime - movingSumRemainingTime / MOVING_SUM_COUNT;
         }
     }
 
+    timeUs_t averageRemainingTime = movingSumRemainingTime / MOVING_SUM_COUNT;
```

Comment:
The comment in the code attempts to explain this.

```
            /* If a task is known to break determinism only schedule it if at least half the average
             * remaining time remains in order to minimise the impact.
             *
             * Also consider the task age, as on startup a task may execute in less than the maxRemainingTime
             * but subsequently the remaining time may never reach that maximum again.
             */
            taskCanBeChosenForScheduling |= (((task->maxExecutionTime > maxRemainingTime) || (task->taskAgeCycles > 1)) &&
            		    					 (remainingTime > averageRemainingTime));
```
The idea here is that if a task could not be scheduled deterministically then it will get scheduled at a time where there will be minimal impact. Of course no task should be taking so long to execute that this expression is required, however this is currently needed to allow tasks such as SERIAL to run without building task specific knowledge into the scheduler. I could use, say, 70% of maxRemainingTime because there's a chance of one rogue, never to be repeated high value, or rather, I can't be sure that there won't be, so basing the above logic on the average time gives a better guarantee that tasks like SERIAL will run OK.